File: share_handler_android/android/src/main/kotlin/com/shoutsocial/share_handler_android/Messages.kt
//package com.shoutsocial.share_handler_android
//
//import android.net.Uri
//import android.util.Log
//import io.flutter.plugin.common.StandardMessageCodec
//import io.flutter.plugin.common.MessageCodec
//import io.flutter.plugin.common.BinaryMessenger
//import io.flutter.plugin.common.BasicMessageChannel
//import java.io.ByteArrayOutputStream
//import java.lang.Error
//import java.lang.RuntimeException
//import java.nio.ByteBuffer
//import java.util.ArrayList
//import java.util.HashMap
//
//private fun wrapError(exception: Throwable): Map<String, Any> {
//    val errorMap: MutableMap<String, Any> = HashMap()
//    errorMap["message"] = exception.toString()
//    errorMap["code"] = exception.javaClass.simpleName
//    errorMap["details"] =
//        "Cause: " + exception.cause + ", Stacktrace: " + Log.getStackTraceString(
//            exception
//        )
//    return errorMap
//}
//
//enum class SharedAttachmentType(val index: Int) {
//    Image(0), Video(1), Audio(2), File(3);
//}
//
//data class SharedAttachment(val path: String, val type: SharedAttachmentType) {
//    fun toMap(): Map<String, Any?> {
//        val toMapResult: MutableMap<String, Any?> = HashMap()
//        toMapResult["path"] = path
//        toMapResult["type"] = type.index
//        return toMapResult
//    }
//
//    companion object {
//        fun fromMap(map: Map<String?, Any?>): SharedAttachment {
//            return SharedAttachment(
//                path = map["path"] as String,
//                type = SharedAttachmentType.values()[map["type"] as Int]
//            )
//        }
//    }
//}
//
//data class SharedMedia(val attachments: List<SharedAttachment>? = null, val conversationIdentifier: String? = null, val content: String? = null, val speakableGroupName: String? = null, val serviceName: String? = null, val senderIdentifier: String? = null, val imageFilePath: String? = null) {
//    fun toMap(): Map<String, Any?> {
//        val toMapResult: MutableMap<String, Any?> = HashMap()
//        toMapResult["attachments"] = attachments?.map { attachment -> attachment.toMap() }
//        toMapResult["conversationIdentifier"] = conversationIdentifier
//        toMapResult["content"] = content
//        toMapResult["speakableGroupName"] = speakableGroupName
//        toMapResult["serviceName"] = serviceName
//        toMapResult["senderIdentifier"] = senderIdentifier
//        toMapResult["imageFilePath"] = imageFilePath
//        return toMapResult
//    }
//
//    companion object {
//        @Suppress("UNCHECKED_CAST")
//        fun fromMap(map: Map<String?, Any?>): SharedMedia {
//            return SharedMedia(attachments = (map["attachments"] as List<*>?)?.map { attachment -> SharedAttachment.fromMap(attachment as Map<String?, Any?>) },
//                conversationIdentifier = map["conversationIdentifier"] as String?,
//                content = map["content"] as String?,
//                speakableGroupName = map["speakableGroupName"] as String?,
//                serviceName = map["serviceName"] as String?,
//                senderIdentifier = map["senderIdentifier"] as String?,
//                imageFilePath = map["imageFilePath"] as String?,
//            )
//        }
//    }
//}
//
//interface Result<T> {
//    fun success(result: T?)
//    fun error(error: Throwable)
//}
//
//private class ShareHandlerApiCodec private constructor() : StandardMessageCodec() {
//    @Suppress("UNCHECKED_CAST")
//    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any {
//        return when (type) {
//            128.toByte() -> SharedAttachment.fromMap(readValue(buffer) as Map<String?, Any?>)
//            129.toByte() -> SharedMedia.fromMap(readValue(buffer) as Map<String?, Any?>)
//            130.toByte() -> SharedMedia.fromMap(readValue(buffer) as Map<String?, Any?>)
//            else -> super.readValueOfType(type, buffer)
//        }
//    }
//
//    override fun writeValue(stream: ByteArrayOutputStream, value: Any) {
//        when (value) {
//            is SharedAttachment -> {
//                stream.write(128)
//                writeValue(stream, value.toMap())
//            }
//            is SharedMedia -> {
//                stream.write(129)
//                writeValue(stream, value.toMap())
//            }
//            is SharedMedia -> {
//                stream.write(130)
//                writeValue(stream, value.toMap())
//            }
//            else -> {
//                super.writeValue(stream, value)
//            }
//        }
//    }
//
//    companion object {
//        val INSTANCE = ShareHandlerApiCodec()
//    }
//}
//
//interface ShareHandlerApi {
//    fun getInitialSharedMedia(result: Result<SharedMedia>?)
//    fun recordSentMessage(media: SharedMedia)
//    fun resetInitialSharedMedia()
//
//    companion object {
//        /** The codec used by ShareHandlerApi.  */
//        private val codec: MessageCodec<Any>
//            get() = ShareHandlerApiCodec.INSTANCE
//
//        @Suppress("UNCHECKED_CAST")
//        fun setup(binaryMessenger: BinaryMessenger, api: ShareHandlerApi?) {
//            if (api == null) return
//            run {
//                val channel = BasicMessageChannel(
//                    binaryMessenger,
//                    "dev.flutter.pigeon.ShareHandlerApi.getInitialSharedMedia",
//                    codec
//                )
//                channel.setMessageHandler { _: Any?, reply: BasicMessageChannel.Reply<Any> ->
//                    val wrapped: MutableMap<String, Any?> = HashMap()
//                    try {
//                        val resultCallback: Result<SharedMedia> =
//                            object : Result<SharedMedia> {
//                                override fun success(result: SharedMedia?) {
//                                    result?.let { wrapped["result"] = result }
//                                    reply.reply(wrapped)
//                                }
//
//                                override fun error(error: Throwable) {
//                                    wrapped["error"] = wrapError(error)
//                                    reply.reply(wrapped)
//                                }
//                            }
//                        api.getInitialSharedMedia(resultCallback)
//                    } catch (exception: Error) {
//                        wrapped["error"] = wrapError(exception)
//                        reply.reply(wrapped)
//                    } catch (exception: RuntimeException) {
//                        wrapped["error"] = wrapError(exception)
//                        reply.reply(wrapped)
//                    }
//                }
//            }
//            run {
//                val channel = BasicMessageChannel(
//                    binaryMessenger,
//                    "dev.flutter.pigeon.ShareHandlerApi.recordSentMessage",
//                    codec
//                )
//                channel.setMessageHandler { message: Any?, reply: BasicMessageChannel.Reply<Any> ->
//                    val wrapped: MutableMap<String, Any?> = HashMap()
//                    try {
//                        val args = message as ArrayList<*>
//                        val mediaArg = args[0]
//                        val media = SharedMedia.fromMap(mediaArg as Map<String?, Any?>)
//                        api.recordSentMessage(media)
//                        wrapped["result"] = null
//                    } catch (exception: Error) {
//                        wrapped["error"] = wrapError(exception)
//                    } catch (exception: RuntimeException) {
//                        wrapped["error"] = wrapError(exception)
//                    }
//                    reply.reply(wrapped)
//                }
//            }
//            run {
//                val channel = BasicMessageChannel(
//                    binaryMessenger,
//                    "dev.flutter.pigeon.ShareHandlerApi.resetInitialSharedMedia",
//                    codec
//                )
//                channel.setMessageHandler { _: Any?, reply: BasicMessageChannel.Reply<Any> ->
//                    val wrapped: MutableMap<String, Any?> = HashMap()
//                    try {
//                        api.resetInitialSharedMedia()
//                        wrapped["result"] = null
//                    } catch (exception: Error) {
//                        wrapped["error"] = wrapError(exception)
//                    } catch (exception: RuntimeException) {
//                        wrapped["error"] = wrapError(exception)
//                    }
//                    reply.reply(wrapped)
//                }
//            }
//        }
//    }
//}

File: share_handler_platform_interface/lib/src/data/messages.dart
import 'dart:async';

import 'package:flutter/foundation.dart' show WriteBuffer, ReadBuffer;
import 'package:flutter/services.dart';

enum SharedAttachmentType {
  image,
  video,
  audio,
  file,
}

class SharedAttachment {
  SharedAttachment({
    required this.path,
    required this.type,
  });

  /// The path to the file on device
  String path;
  SharedAttachmentType type;

  Object encode() {
    final Map<Object?, Object?> pigeonMap = <Object?, Object?>{};
    pigeonMap['path'] = path;
    pigeonMap['type'] = type.index;
    return pigeonMap;
  }

  static SharedAttachment decode(Object message) {
    final Map<Object?, Object?> pigeonMap = message as Map<Object?, Object?>;
    return SharedAttachment(
      path: Uri.decodeFull(pigeonMap['path']! as String),
      type: SharedAttachmentType.values[pigeonMap['type']! as int],
    );
  }
}

class SharedMedia {
  SharedMedia({
    this.attachments,
    this.recipientIdentifiers,
    this.conversationIdentifier,
    this.content,
    this.speakableGroupName,
    this.serviceName,
    this.senderIdentifier,
    this.imageFilePath,
  });

  /// List of shared attachments (ex. images, videos, pdfs, etc.). Each attachment has an attachment type and a path to the file on the device.
  List<SharedAttachment?>? attachments;

  /// iOS only: List of recipient identifiers from iOS intent.
  List<String?>? recipientIdentifiers;

  /// The identifier of the conversation that content was shared to. This will come back if you use the 'recordSentMessage' method, and the user selects a specific conversation to share content to.
  String? conversationIdentifier;

  /// Text content that was shared if any. Could be a url as well.
  String? content;

  /// The name of the recipient the content was shared to if specified.
  String? speakableGroupName;

  /// iOS only: The name of the service that sent the content.
  String? serviceName;

  /// iOS only: The identifier of the sender that shared the content.
  String? senderIdentifier;

  /// iOS only: The file path for the image of the sender.
  String? imageFilePath;

  Object encode() {
    final Map<Object?, Object?> pigeonMap = <Object?, Object?>{};
    pigeonMap['attachments'] = attachments;
    pigeonMap['recipientIdentifiers'] = recipientIdentifiers;
    pigeonMap['conversationIdentifier'] = conversationIdentifier;
    pigeonMap['content'] = content;
    pigeonMap['speakableGroupName'] = speakableGroupName;
    pigeonMap['serviceName'] = serviceName;
    pigeonMap['senderIdentifier'] = senderIdentifier;
    pigeonMap['imageFilePath'] = imageFilePath;
    return pigeonMap;
  }

  static SharedMedia decode(Object message) {
    final Map<Object?, Object?> pigeonMap = message as Map<Object?, Object?>;
    return SharedMedia(
      attachments: (pigeonMap['attachments'] as List<Object?>?)
          ?.map((e) => SharedAttachment.decode(e as Map<Object?, Object?>))
          .cast<SharedAttachment?>()
          .toList(),
      recipientIdentifiers: (pigeonMap['recipientIdentifiers'] as List<Object?>?)?.cast<String?>(),
      conversationIdentifier: pigeonMap['conversationIdentifier'] as String?,
      content: pigeonMap['content'] as String?,
      speakableGroupName: pigeonMap['speakableGroupName'] as String?,
      serviceName: pigeonMap['serviceName'] as String?,
      senderIdentifier: pigeonMap['senderIdentifier'] as String?,
      imageFilePath: pigeonMap['imageFilePath'] as String?,
    );
  }
}

class _ShareHandlerApiCodec extends StandardMessageCodec {
  const _ShareHandlerApiCodec();
  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    if (value is SharedAttachment) {
      buffer.putUint8(128);
      writeValue(buffer, value.encode());
    } else if (value is SharedMedia) {
      buffer.putUint8(129);
      writeValue(buffer, value.encode());
    } else if (value is SharedMedia) {
      buffer.putUint8(130);
      writeValue(buffer, value.encode());
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    switch (type) {
      case 128:
        return SharedAttachment.decode(readValue(buffer)!);

      case 129:
        return SharedMedia.decode(readValue(buffer)!);

      case 130:
        return SharedMedia.decode(readValue(buffer)!);

      default:
        return super.readValueOfType(type, buffer);
    }
  }
}

class ShareHandlerApi {
  /// Constructor for [ShareHandlerApi].  The [binaryMessenger] named argument is
  /// available for dependency injection.  If it is left null, the default
  /// BinaryMessenger will be used which routes to the host platform.
  ShareHandlerApi({BinaryMessenger? binaryMessenger}) : _binaryMessenger = binaryMessenger;

  final BinaryMessenger? _binaryMessenger;

  static const MessageCodec<Object?> codec = _ShareHandlerApiCodec();

  Future<SharedMedia?> getInitialSharedMedia() async {
    final BasicMessageChannel<Object?> channel = BasicMessageChannel<Object?>(
        'dev.flutter.pigeon.ShareHandlerApi.getInitialSharedMedia', codec,
        binaryMessenger: _binaryMessenger);
    final Map<Object?, Object?>? replyMap = await channel.send(null) as Map<Object?, Object?>?;
    if (replyMap == null) {
      throw PlatformException(
        code: 'channel-error',
        message: 'Unable to establish connection on channel.',
      );
    } else if (replyMap['error'] != null) {
      final Map<Object?, Object?> error = (replyMap['error'] as Map<Object?, Object?>?)!;
      throw PlatformException(
        code: (error['code'] as String?)!,
        message: error['message'] as String?,
        details: error['details'],
      );
    } else if (replyMap['result'] != null) {
      if (replyMap['result'] is SharedMedia) {
        return replyMap['result'] as SharedMedia;
      }

      return SharedMedia.decode(replyMap['result']!);
    } else {
      return null;
    }
  }

  Future<void> recordSentMessage(SharedMedia argMedia) async {
    final BasicMessageChannel<Object?> channel = BasicMessageChannel<Object?>(
        'dev.flutter.pigeon.ShareHandlerApi.recordSentMessage', codec,
        binaryMessenger: _binaryMessenger);
    final Map<Object?, Object?>? replyMap = await channel.send(<Object?>[argMedia]) as Map<Object?, Object?>?;
    if (replyMap == null) {
      throw PlatformException(
        code: 'channel-error',
        message: 'Unable to establish connection on channel.',
      );
    } else if (replyMap['error'] != null) {
      final Map<Object?, Object?> error = (replyMap['error'] as Map<Object?, Object?>?)!;
      throw PlatformException(
        code: (error['code'] as String?)!,
        message: error['message'] as String?,
        details: error['details'],
      );
    } else {
      return;
    }
  }

  Future<void> resetInitialSharedMedia() async {
    final BasicMessageChannel<Object?> channel = BasicMessageChannel<Object?>(
        'dev.flutter.pigeon.ShareHandlerApi.resetInitialSharedMedia', codec,
        binaryMessenger: _binaryMessenger);
    final Map<Object?, Object?>? replyMap = await channel.send(null) as Map<Object?, Object?>?;
    if (replyMap == null) {
      throw PlatformException(
        code: 'channel-error',
        message: 'Unable to establish connection on channel.',
      );
    } else if (replyMap['error'] != null) {
      final Map<Object?, Object?> error = (replyMap['error'] as Map<Object?, Object?>?)!;
      throw PlatformException(
        code: (error['code'] as String?)!,
        message: error['message'] as String?,
        details: error['details'],
      );
    } else {
      return;
    }
  }
}
